pysoa/test/plan/parser.py

```diff
@@ -0,0 +1,289 @@
+from __future__ import absolute_import, unicode_literals
+
+import codecs
+import copy
+import functools
+import os
+
+from pyparsing import (
+    col as get_parse_column,
+    line as get_parse_line,
+    lineno as get_parse_line_number,
+    LineEnd,
+    LineStart,
+    MatchFirst,
+    OneOrMore,
+    ParseException,
+    StringEnd,
+    StringStart,
+)
+
+from pysoa.test.plan.errors import (
+    FixtureLoadError,
+    FixtureSyntaxError,
+)
+from pysoa.test.plan.grammar.data_types import DataTypeConversionError
+from pysoa.test.plan.grammar.directive import get_all_directives
+from pysoa.test.plan.grammar.tools import (
+    path_get,
+    path_put,
+    get_all_paths,
+)
+
+
+class ServiceTestPlanFixtureParser(object):
+    """
+    This class is responsible for parsing test plans. Create an instance using a fixture file name and fixture name,
+    and then call ``parse_test_fixture`` to parse the fixture file and return all the test cases in it.
+    """
+
+    def __init__(self, fixture_file_name, fixture_name):
+        """
+        Construct a test fixture parser.
+
+        :param fixture_file_name: The path to and name of the fixture file
+        :type fixture_file_name: union[str, unicode]
+        :param fixture_name: The name of the fixture
+        :type fixture_name: union[str, unicode]
+        """
+        self._fixture_file_name = fixture_file_name
+        self._fixture_name = fixture_name
+        self._working_directory = os.path.basename(fixture_file_name)
+
+        if not os.path.isabs(self._fixture_file_name):
+            self._fixture_file_name = os.path.abspath(os.path.join(self._working_directory, self._fixture_file_name))
+
+        self.test_cases = []
+        self._global_directives = {}
+        self._working_action_case = {}
+        self._working_test_case = {}
+        self._working_test_case_line_number = 0
+        self._working_test_case_source = []
+        self._fixture_source = []
+        self._grammar = None
+        self._line_number = 0
+
+    def parse_test_fixture(self):
+        """
+        The main method responsible for the start-to-finish process of parsing test cases from a fixture file. It uses
+        the other methods on this class, which generally should not be used directly.
+
+        :return: The list of test cases
+        :rtype: list[dict]
+        """
+        self._grammar = self._compile_grammar()
+
+        self._ingest_directives()
+
+        if not self.test_cases:
+            raise FixtureLoadError('Failed to find any test cases in file: {}'.format(self._fixture_file_name))
+
+        return self.test_cases
+
+    def _ingest_directives(self):
+        try:
+            with codecs.open(self._fixture_file_name, mode='rb', encoding='utf-8') as file_input:
```

Comment:
Why use `codecs.open`? Should we use `io.open` instead?

Reply:
A while back, I did a bunch of research about what is the best way to open a file with an encoding that would work exactly the same in both Python 2 and Python 3, and the majority opinion leaned heavily toward `codecs.open` (since straight `open` doesn't support an encoding until Python 3.x). As such, that's the way I have done it in all of our Python 2- and 3-compatible code.